### ISSUE TITLE
Make percentile unique

### ISSUE THREAD

**Issue by dcesari** [ARPA-SIMC/arkimet#201]
In the Mistral project we will need to archive grib messages edition 2 containing percentile values, i.e. using templates 4.6 or 4.10 and differing only by `percentileValue` key, apart from values. This means that we need to make them unique, if requested, according to that key in the scanning phase.
Would this provide also a query key for free (e.g. in proddef) or that would require additional work?

Attached is an example, depleted of data, with a set of percentiles to show the issue:
[percentile.zip](https://github.com/ARPA-SIMC/arkimet/files/3751451/percentile.zip).

Something similar may be needed also for probability forecasts if not already implemented (templates 4.5 and 4.9), but we will speak about that at next meeting.

**Comment by edigiacomo** [comment]
I'd use the `proddef` type for `percentileValue` - and it should be very simple to implement.

For the probability forecast, do you have some examples and/or the GRIB key name?

**Comment by dcesari** [comment]
I did not want to put too much meat on fire :-), but if you insist...
```
37        probabilityType = 3 [Probability of event above lower limit  (grib2/tables/11/4.9.table) ]
38        scaleFactorOfLowerLimit = -1
39-42     scaledValueOfLowerLimit = 3
43        scaleFactorOfUpperLimit = 0
44-47     scaledValueOfUpperLimit = 0
```
if ANY of these keys is different it is a different probability product. I can provide an example tomorrow.

**Comment by edigiacomo** [comment]
`proddef` keys in Arkimet are represented with 1-3 chars. These are the current available values:

* `tod: typeOfProcessedData`
* `mc: marsClass`
* `mt: marsType`
* `ty: typeOfEnsembleForecast`
* `pf: perturbationNumber`
* `tf: numberOfForecastsInEnsemble`
* `d: derivedForecast`

Following this style, we could add:

* `pv: percentileValue`
* `pt: probabilityType`
* `fll : scaledFactorOfLowerLimit`
* `vll: scaledValueOfLowerLimit`
* `ful: scaledFactorOfUpperLimit`
* `vul: scaledValueOfUpperLimit`

Or we could create a comma separated string for the scaled factors `pl: scaledFactorOfLowerLimit,scaledValueOfLowerLimit,scaledFactorOfUpperLimit,scaledValueOfUpperLimit` (e.g. `proddef: GRIB:pt=3,pl=-1,3,0,0`) but in this case you can't query the scaled factors individually: `proddef: GRIB:pl=-1,3,0,0` is a valid query but `proddef:GRIB:pl=-1,,0,` is invalid.

I have implemented the latter in branch `issue201`.

**Comment by dcesari** [comment]
What a rocket!
Well, for percentile, my guess is that the most common query, apart from "all the percentile values", would be to ask for an interval v1<=pv<=v2 which is probably not straightforward to implement, so feel free to refuse.
For probability, different combinations of scaledValue and scaleFactor (notice the 'd' only before Value) could give the same actual result (e.g. 1,0 or 100,2 both =1.), so this could disturb the sleep of sensitive people like us :-). I think the typical use case is to ask for all the probabilities of a given variable, so I would go for the simplest and lightest solution, probably the one you already implemented, with a single key `pl` encompassing all the 4 grib keys.
What do people in Reading think?

**Comment by edigiacomo** [comment]
> Well, for percentile, my guess is that the most common query, apart from "all the percentile values", would be to ask for an interval v1<=pv<=v2 which is probably not straightforward to implement, so feel free to refuse.

TL;DR: it's not easy to implement :)

`proddef` has the same limitations of `timerange` and `level`: if you want to ask for an interval, you must list all the values in the interval, i.e. `proddef:GRIB:pv=v1 or GRIB:pv=.. or GRIB:pv=v2`. 

We could implement a new type similar to `reftime` (the only type that support intervals) or extend the behaviour of the `ValueBag` with `>, >=, <=, <,∧` operators (e.g. `proddef:GRIB:pv>=v1,GRIB:pv<=2`), but then I don't know how to we could combine the `and` and the `or` operator in more complex queries. Otherwise, you can parse the JSON summary, match all the `pv` in the interval and then expand the list (as you can do with `timerange` and `level`).

> I would go for the simplest and lightest solution, probably the one you already implemented, with a single key `pl` encompassing all the 4 grib keys.

Yes, I already have implemented this solution. You can test it in branch `issue201`:

```
$ cd /path/to/arkimet
$ git checkout -t origin/issue201
$ make
$ ./run-local src/arki-scan --dump FILE.GRIB
```

If it's good, I'll merge in the `issue177` branch.

**Comment by spanezz** [comment]
Extending matching of ValueBag would totally be possible. Open a bug if you decide it could be helpful, and I'll be very happy to draft a syntax for it and implement it.

However, keep in mind that arkimet works better if number of different metadata values stays low (excluding reftime which is indexed differently), so be careful not to make it so that having the possibility of matching ranges of ValueBag elements doesn't encourage fine-grained, never-equal values stored in proddef

**Comment by dcesari** [comment]
The current implementation is satisfactory, thanks.

**Comment by edigiacomo** [comment]
I reopened the issue, because the some tests fails 

```
AssertionError: 'GRIB(pl=None,3,None,0, pt=3, tod=5)' != 'GRIB(tod=5)'
- GRIB(pl=None,3,None,0, pt=3, tod=5)
+ GRIB(tod=5)
```

I think it's just a typo in the GRIB keys. I'll open a PR when fixed.
